awslabs/aws-lambda-powertools-typescript
# feat: added captureHTTPsRequest feature

## Description of your changes

This PR introduces a new feature for Tracer that allows to trace, by default, all outgoing HTTP(s) requests:
- Customers can opt-out from this feature by setting the `POWERTOOLS_TRACER_CAPTURE_HTTPS_REQUESTS` environment variable or `captureHTTPsRequests` constructor parameter to false when instantiating the Tracer
- The feature relies on `aws-xray-sdk` ability to instrument the standard library's `http` and `https` modules, as Tracer should be able to trace any 3rd party request client library that is built upon these two modules. Support to 3rd party HTTP clients is provided on a best effort basis (this disclaimer is specified in the documentation). `axios` has been tested during development and is included in integration tests.
- Documentation and JSDocs updated
- The feature is implemented without exposing any new API on Tracer and directly on the Tracer instance, as such no changes to decorator and middleware implementation has been made
- Unit tests implemented
- Integration tests implemented
- Minor housekeeping changes (mostly formatting)

### How to verify this change

See checks passing on this PR, see integration test results ([link](https://github.com/awslabs/aws-lambda-powertools-typescript/runs/5669680344?check_suite_focus=true))

Screenshots of the new section in the documentation:
![image](https://user-images.githubusercontent.com/7353869/159504583-56dffe8c-2a34-4b93-9373-e778e8ba8767.png)

Documentation changes also include the addition of the new config to the Utility Settings section:
![image](https://user-images.githubusercontent.com/7353869/159504725-502b4939-5953-4266-913b-7cbe62329013.png)

Screenshots of traces & service map generated by running this build & using `axios`:
![Screenshot 2022-03-22 at 14 37 02](https://user-images.githubusercontent.com/7353869/159502965-da8f881c-ae51-43d9-8de7-26db66d7ce3b.png)
![Screenshot 2022-03-22 at 14 36 45](https://user-images.githubusercontent.com/7353869/159503003-fec16635-0d9a-4a1f-8648-556c2913454b.png)


### Related issues, RFCs

[#675](https://github.com/awslabs/aws-lambda-powertools-typescript/issues/675)  
[#275](https://github.com/awslabs/aws-lambda-powertools-typescript/issues/275)

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist


### Breaking change checklist

N/A

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
